src/Compilers/Core/Portable/PEWriter/MetadataWriter.PortablePdb.cs

```diff
@@ -745,8 +745,14 @@ private DocumentHandle AddDocument(DebugSourceDocument document, Dictionary<Debu
             DocumentHandle documentHandle;
             DebugSourceInfo info = document.GetSourceInfo();
 
+            var name = document.Location;
+            if (_sourceDebugDocumentsAdded && this.Context.RebuildDataOpt is { } rebuildData)
+            {
+                name = rebuildData.GetNextNonSourceFileDocumentName();
```

Comment:
>name = rebuildData.GetNextNonSourceFileDocumentName() [](http://example.com/codeflow?start=16&length=53)

It looks fragile to rely on the order that `AddDocument()` is called. Could we simply add all `NonSourceFileDocumentNames` directly at the point in `MetadataWriter.BuildMetadataAndIL()` where we set `_sourceDebugDocumentsAdded`?

Reply:
Unfortunately no we can't. 

Have to get into how `#line` are processed a bit to understand the problem here. The `#line` directives are processed initially in `GenerateMethodBody` and that runs in parallel across the compilation. As they are processed there we essentially have to do two operations: 

- Generate a path via `NormalizePath`. The invariant here is that for two calls to `NormalizePath` for the same logical file name in a `#line` directive we need to return the same string. By "logical name" I mean that two files in the same directive with a `#line 42 "data.txt"` need to return the same name but a file in a different directory with the same `#line 42 "data.txt"` needs to return a different name. 
- Generate a `DebugSourceDocument` for that path. As long as we maintain the invariant above this will just work out. 

This gets us up to the point where we set `_sourceDebugDocumentsAdded` to `true`. At this point we have:

1. The dictionary of `string` (the path) to the `DebugSourceDocument` entries 
2. The set of non-source file names we need to emit into the PDB 

While we absolutely know the order in which (2) needs to be emitted we don't know the order of (1) and that's important. The final stage of emit is going to lookup `DebugSourceDocument` in the same order as the original compilation via `AddDocument` and we have to make sure all the identities match here (particularly the numeric token attached with the names, documents, etc ...) It's also not possible to map (2) back to (1) without a decent investment in how we deal with full paths. 

In chatting with @tmat we favored just hooking `AddDocument` since because of determinism we understood that call will come in a predictable order. 

I think with enough work I could make it possible to do the name mapping correctly in `NormalizePath` but I worry I'm going to miss alot of casese. Particularly if users invest in very non-standard pathing conventions in `/pathmap`. 

Very open to suggestions here if you see a better way.